### ISSUE TITLE
wrong PushButton color

### ISSUE THREAD

**Issue by pyxiscloud** [macosui/macos_ui#486]
**Description**
After update to 2.0.2, PushButton colors always blue, even if I set it to another color

**Steps To Reproduce**

<details>
<summary>PushButton(
          controlSize: ControlSize.large,
          color: Colors.redAccent, 
)</summary>

```[tasklist]
### Tasks
```

```[tasklist]
### Tasks
```

**Comment by GroovinChip** [comment]
Thanks for reporting this bug @pyxiscloud. I've reproduced it.

@Adrian-Samoticha could this be related to the changes we shipped for `PushButton` in version 2.0.1? We listen for the system accent color now, but I don't recall if that change means we disregard the `color` property now. I don't think that's the case, because we have this function:
![_getBackgroundColor](https://github.com/macosui/macos_ui/assets/4250470/613dffa5-315e-4a1c-869c-068d35128d8b)

As highlighted, line 285 shows that we default to the `color` property. 

I observed that we were not adding the `backgroundColor` to our `BoxDecoration`, as highlighted on line 360 below, but that did not resolve the issue. 

![building_the_button](https://github.com/macosui/macos_ui/assets/4250470/dc3d0045-4899-4113-bf12-f11579d46dea)

However, adding it to the `Container` _does_ result in the button being the specified color; unfortunately, it also results in the rest of the button's styling being completely ignored:
<img width="912" alt="Screenshot 2023-10-15 at 2 06 20 PM" src="https://github.com/macosui/macos_ui/assets/4250470/4dcf1032-960a-4b62-bc58-e5006d6aa184">

I am stumped.

**Comment by Adrian-Samoticha** [comment]
That’s interesting. I’ll try to reproduce it and see what I can find out.

**EDIT:** The container exists solely to apply the “click effect” decoration because the `Container` widget offers a `foregroundDecoration` property and that decoration needed to be in the foreground. It makes sense that adding a color to it would hide the decoration of its parent `DecoratedBox`.

`_getBackgroundColor` seems to only be used to determine the foreground color (the text is either white or black depending on the brightness of the background color), so that explains why the color doesn’t show up.

The reason why adding the background color to the `BoxDecoration` doesn’t do anything is that it is styled using a gradient now, rather than a singular color. 

You can assign me to that issue, I think I know how to fix it.

---

That said, @pyxiscloud, if the reason for why you would like to change the background color of the button is that you would like to style the entire application in that particular color, the proper way to do so is like this:

Open the `macos/Runner.xcworkspace` folder of your project using Xcode and search for the `Assets.xcassets` file. Then, hit the plus-button and choose “Color Set”:

<img width="1410" alt="image" src="https://github.com/macosui/macos_ui/assets/86920182/005f5861-bdc3-4a28-a687-0328778050b2">

Rename the newly created color set to `AccentColor` and choose your preferred color:


<img width="1512" alt="Screenshot 2023-10-16 at 14 14 46" src="https://github.com/macosui/macos_ui/assets/86920182/a17c6fd1-2583-436e-9dfa-dd249a4c939b">


Then, go to `Runner.xcodeproj` > `Build Settings` and under “Asset Catalog Compiler – Options” locate the “Global Accent Color Name” property, and set it to `AccentColor`:

<img width="1512" alt="Screenshot 2023-10-16 at 14 11 44" src="https://github.com/macosui/macos_ui/assets/86920182/d496de9e-cf2d-4e17-9d94-b590dfcacfdb">


Violà:


<img width="912" alt="Screenshot 2023-10-16 at 14 14 15" src="https://github.com/macosui/macos_ui/assets/86920182/104d276f-d03f-4723-9966-1c9417b3b3ad">


The benefit of this approach is that, for one, macos_ui will choose a gradient that matches those of native macOS applications, and for two, the user will still be able to override the color via the System Settings, which is how native macOS applications behave.

I guess it wouldn’t hurt to add this guide to macos_ui’s readme, come to think of it.

**Comment by pyxiscloud** [comment]
will "color" be deprecated for PushButton?

**Comment by kwang87** [comment]
I used a pushbutton to remove the background color and use it like a text button. Is this impossible now?

**Comment by Adrian-Samoticha** [comment]
> will "color" be deprecated for PushButton?

No, but if you’re trying to style the entire application, rather than just a single button, the way I described is the way to do so. The fact that the `color` property doesn’t change the color of the button is, indeed, a bug that will need to be fixed.

**Comment by Adrian-Samoticha** [comment]
> I used a pushbutton to remove the background color and use it like a text button. Is this impossible now?

That’s possible, but this is considered a bug, so it will be fixed.

**EDIT:** That said, I am not aware of any native text buttons. Are you trying to emulate a specific macOS UI component that is missing in macos_ui?

**Comment by ralph-bergmann** [comment]
The [workaround](https://github.com/macosui/macos_ui/issues/486#issuecomment-1764197632) doesn't work for all buttons/UI elements for me :-(

<img width="1738" alt="SCR-20231207-mbes" src="https://github.com/macosui/macos_ui/assets/1191649/18886601-3ed1-46f2-b451-6253f2a7302d">

I wonder why it doesn't just use the `MacosThemeData` `primaryColor` for this? Isn't that what the accent color is for?

On the other hand, the workaround fixes #395.

**Comment by Adrian-Samoticha** [comment]
> The [workaround](https://github.com/macosui/macos_ui/issues/486#issuecomment-1764197632) doesn't work for all buttons/UI elements for me :-(
> 
> —
> 
> On the other hand, the workaround fixes #395.

Currently, only the buttons support accent colors. #484 introduces accent color support for the sidebar items but has not yet been merged. The other widgets are, of course, planned, but development is rather slow right now.

**Comment by Adrian-Samoticha** [comment]
> I wonder why it doesn't just use the `MacosThemeData` `primaryColor` for this? Isn't that what the accent color is for?

I tried applying the colors provided by [appkit_ui_element_colors](https://pub.dev/packages/appkit_ui_element_colors) to the widgets directly, however, this didn’t make them look accurate to their native counterparts. It’s necessary to overhaul each widget individually to make it look good.

**Comment by Adrian-Samoticha** [comment]
@GroovinChip While #484 is somewhat related to this issue, it doesn’t actually resolve it. I’ll quickly reopen this issue and see if I’ll have time to work on it sometime.

**Comment by GroovinChip** [comment]
Ah yeah, you're right. My bad. Thanks for catching this!

**Comment by ralph-bergmann** [comment]
I don't know if each widget should read the color from the system itself. Then, the theme's primary color becomes useless. If a user of this library wants to overwrite the color, he has to do this manually for each widget, as the theme no longer supports it (the `PushButtonTheme` is already deprecated).

IMHO, this should be done by the theme (`primaryColor ?== readSystemnAccentColor()`), and all widgets should use the primaryColor from the theme.

**Comment by Adrian-Samoticha** [comment]
> I don't know if each widget should read the color from the system itself. Then, the theme's primary color becomes useless. If a user of this library wants to overwrite the color, he has to do this manually for each widget, as the theme no longer supports it (the `PushButtonTheme` is already deprecated).
> 
> IMHO, this should be done by the theme (`primaryColor ?== readSystemnAccentColor()`), and all widgets should use the primaryColor from the theme.

In order to make the macos_ui components’ appearance as accurate to their native counterparts as possible, the supported colors currently use hard-coded gradients for each system accent color setting, rather than being generated from a single color. Therefore, that wouldn’t really be possible without sacrificing accuracy to macOS’ native UI components.

The intended way to theme the entire application is the one I’ve outlined above, and I am planning to add this guide to macos_ui’s readme, once all widget actually support system accent colors.

**Comment by Adrian-Samoticha** [comment]
> The [workaround](https://github.com/macosui/macos_ui/issues/486#issuecomment-1764197632) doesn't work for all buttons/UI elements for me :-(

I know it’s been a while, but version 2.0.7 applies the accent color to most widgets now.

**Comment by akhudek** [comment]
My use case for this is to make destructive buttons like delete red. Is there another way of doing this?

**Comment by Adrian-Samoticha** [comment]
> My use case for this is to make destructive buttons like delete red. Is there another way of doing this?

Apple likes to style these buttons by making the text red, rather than the button’s background:

<img width="322" alt="image" src="https://github.com/macosui/macos_ui/assets/86920182/48b93aca-4583-4c7c-81c9-5d751ca0e004">

Since the `PushButton` class takes a widget as a child, you could probably just supply it with a red `Text` widget. Be sure to set the `secondary` property of the `PushButton` to true to prevent it from being colored and I think you’re good to go.
